includes/class-wcs-cart-renewal.php

```diff
@@ -1609,26 +1609,16 @@ public function set_renewal_order_cart_hash_on_block_checkout( $has_status, $ord
 		 * this is the status check in DraftOrderTrait::is_valid_draft_order()) and the order doesn't have that status. Orders
 		 * which already have the checkout-draft status don't need to be updated to bypass the checkout block logic.
 		 */
-		if ( $has_status || 'checkout-draft' !== $status ) {
+		if ( $has_status || ! $order || 'checkout-draft' !== $status ) {
 			return $has_status;
 		}
 
-		/**
-		 * This function is only concerned with updating the order cart hash during REST API requests - which is the request
-		 * context where the Store API Checkout Block validates the order for payment resumption.
-		 */
-		if ( ! WC()->is_rest_api_request() ) {
-			return $has_status;
-		}
+		$cart_order = $this->get_order();
 
 		// If the order being validated is the order in the cart, then we need to update the cart hash so it can be resumed.
-		if ( $order && $order->get_id() === WC()->session->get( 'store_api_draft_order', 0 ) ) {
```

Comment:
> However, this effectively means that any has_status( 'checkout-draft' ) will attempt to check if the cart contains a renewal order that matches the order being checked.
>
> I'm sure how we can make this logic tighter.

@james-allan Looking over the stacktrace you put in Slack and also at the [`DraftOrderTrait` class](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/Utilities/DraftOrderTrait.php) in Woo Blocks, it looks like we should be able to bring back the `wc()->session->get( 'store_api_draft_order', 0 )` condition check because this order ID stored in the session should always exist in order for `is_valid_draft_order()` to be calling `has_status( 'checkout-draft' )`, which we're hooking on to.

Also, we chatted about this outside of GH but if there are concerns regarding performance, another small change we could make is with `$cart_order = $this->get_order();`.

Since we only care about the renewal order ID, we could refactor the `WCS_Cart_Renewal::get_order()` function to have a new `$returns = object|id` arg (defaulted to 'object').
I think you mentioned that this class is a parent of a few similar cart classes so it would require changing a bunch of functions. From a quick look, it looks like it would be changes across both Woo Subscriptions and subscriptions:

- WCS_Cart_Early_Renewal
- WCS_Cart_Initial_Payment
- WCS_Cart_Renewal
- WCS_Cart_Resubscribe

If we can bring back the check for `store_api_draft_order` from session that'll probably be enough IMO

Reply:
Thanks of the feedback @mattallan. 

>  looks like we should be able to bring back the wc()->session->get( 'store_api_draft_order', 0 ) condition check

When I was working on this fix, I noticed that condition was also failing in the flow I was testing so I got rid of it originally. After observing what the `store_api_draft_order` session value is, I realise it was a string value of the order ID not and int and so the strict `$order->get_id() === wc()->session->get( 'store_api_draft_order' )` didn't pass. 

I've fixed that by casting it to an int. 

> If we can bring back the check for store_api_draft_order from session that'll probably be enough IMO

Yeah I agree, I don't think that's necessary to refactor `get_order()` given we're already pretty restrictive in when we're getting down to that point and at the end of the day it's just a single instance of loading an order so it can't be too performance heavy.